File: indexer/walker.ts
import * as fs from "fs";
import * as path from "path";

// Folder names we NEVER want to walk into.
// These either aren't "your code" (dependencies) or aren't useful for RAG (build output, git internals).
const SKIP_DIRS = new Set([
  "node_modules",
  ".git",
  "dist",
  "build",
  ".next",
  "coverage",
]);

// File types we know how to chunk meaningfully right now.
// We'll add more languages later — starting small on purpose.
const SUPPORTED_EXTENSIONS = new Set([".js", ".ts", ".jsx", ".tsx"]);

// Anything bigger than this is probably a generated/minified file, not real hand-written code.
const MAX_FILE_SIZE_BYTES = 500_000; // ~500kb

export interface WalkResult {
  filePath: string; // full path on disk
  relativePath: string; // path relative to the repo root, easier to read/display
}

/**
 * Walks a folder recursively and returns the list of files we should index.
 * Skips junk folders and unsupported/oversized files.
 */
export function walkRepo(rootDir: string): WalkResult[] {
  const results: WalkResult[] = [];

  function walk(currentDir: string) {
    const entries = fs.readdirSync(currentDir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(currentDir, entry.name);

      if (entry.isDirectory()) {
        if (SKIP_DIRS.has(entry.name)) {
          continue; // skip this folder entirely, don't even look inside
        }
        walk(fullPath); // recurse into subfolder
        continue;
      }

      // it's a file — check if we care about it
      const ext = path.extname(entry.name);
      if (!SUPPORTED_EXTENSIONS.has(ext)) {
        continue;
      }

      const stats = fs.statSync(fullPath);
      if (stats.size > MAX_FILE_SIZE_BYTES) {
        continue; // too big, probably not hand-written code
      }

      results.push({
        filePath: fullPath,
        relativePath: path.relative(rootDir, fullPath),
      });
    }
  }

  walk(rootDir);
  return results;
}

File: indexer/walker.py
import os
from chunker import EXTENSION_TO_LANGUAGE

# Folder names we NEVER want to walk into.
# These either aren't "your code" (dependencies) or aren't useful for RAG (build output, git internals).
SKIP_DIRS = {
    "node_modules", ".git", "dist", "build", ".next", "coverage",
    "target", "__pycache__", ".venv", "venv", "vendor", "bin", "obj",
}

# File types we know how to chunk meaningfully right now. Derived from
# chunker.py's EXTENSION_TO_LANGUAGE so this list can never silently
# drift out of sync with what the chunker actually supports (this used
# to happen: .ts was "supported" here but parsed with the wrong grammar).
SUPPORTED_EXTENSIONS = set(EXTENSION_TO_LANGUAGE.keys())

# Anything bigger than this is probably a generated/minified file, not real hand-written code.
MAX_FILE_SIZE_BYTES = 500_000  # ~500kb


def walk_repo(root_dir):
    """
    Walks a folder recursively and returns the list of files we should index.
    Skips junk folders and unsupported/oversized files.
    Returns a list of dicts: {"file_path": ..., "relative_path": ...}
    """
    results = []

    for current_dir, dir_names, file_names in os.walk(root_dir):
        # Modify dir_names IN PLACE to stop os.walk from ever entering skipped folders.
        dir_names[:] = [d for d in dir_names if d not in SKIP_DIRS]

        for file_name in file_names:
            ext = os.path.splitext(file_name)[1]
            if ext not in SUPPORTED_EXTENSIONS:
                continue

            full_path = os.path.join(current_dir, file_name)

            if os.path.getsize(full_path) > MAX_FILE_SIZE_BYTES:
                continue  # too big, probably not hand-written code

            relative_path = os.path.relpath(full_path, root_dir)

            results.append({
                "file_path": full_path,
                "relative_path": relative_path,
            })

    return results